main.cpp

```diff
@@ -228,6 +229,13 @@ float submit_performance_test(std::string& message, unsigned num_batches,
             throw std::runtime_error("Allocation failed");
     }
 
+    for (auto &bo : cmdbufs) {
+        bo = new GemBuffer(drm);
+
+        if (bo->allocate(4096))
+            throw std::runtime_error("Allocation failed");
+    }
+
```

Comment:
I'd prefer if the cmdbufs and relocs vectors would be just std::vector<GemBuffer> and not deal with pointers. If we add a move constructor like
```
GemBuffer::GemBuffer(GemBuffer &&other)
: _dev(other._dev), _valid(other._valid), _handle(other._handle)
, _size(other._size), _map(other._map)
{
    other._map = nullptr;
    other._valid = false;
}
```
then we can use emplace_back(std::move(bo)) to initialize the bo's in a loop. This way we don't need to worry about freeing the bo's -- here the cmdbuf bo's are never freed, and the reloc bo's are not freed if there's an exception.

Reply:
Wouldn't be cleaner to add something like `->setDRM()` to GemBuffer?

IIUC, with your approach we'd have to create a dummy BO, then clone it by putting to the vector and then iterate through the vector allocating BO's.

Reply:
Yeah, that works too.

With my above approach we would have something like
```
std::vector<GemBuffer> bos;
for (size_t i = 0; i < num_cmdbufs; i++) {
  GemBuffer buf(drm);
  if (buf.allocate(4096)) throw ...;
  bos.emplace_back(std::move(buf));
}
```

Either seems OK.

Reply:
Does `std::move` prevents buf's destructor invocation? At quick glance at the man `std::move` only tries to perform 'efficient' copy.

Reply:
No, but the move constructor (in my first comment) will set _valid to false and _map to nullptr which will make the destructor a no-op.

Reply:
Aha, right. I'll go with `setDRM` if you don't mind, less tangled to me and copy-less.

Reply:
Sure. C++'s move semantics are really complicated.. Rust is so much nicer here ;)